From: Eli Hersha <15244329-BeastRein@users.noreply.gitlab.com>
Date: Thu, 3 Aug 2023 19:12:37 -0400
Subject: [PATCH] drm/panel: add sofef00 panel for bonito

The driver name was changed to include the Pixel 3a XL codename so it
wouldn't conflict with the OnePlus 6 panel driver.

[richard: add kconfig, makefile, commit message; change dts compat string]
[richard: reformat top comment; split prepare/enable and unprepare/disable]
Signed-off-by: Richard Acayan <mailingradian@gmail.com>
---
 drivers/gpu/drm/panel/Kconfig                      |   9 +
 drivers/gpu/drm/panel/Makefile                     |   1 +
 ...ng-s6e3fa7.c => panel-samsung-sofef00-bonito.c} | 185 ++++++++++++---------
 3 files changed, 114 insertions(+), 81 deletions(-)
 copy drivers/gpu/drm/panel/{panel-samsung-s6e3fa7.c => panel-samsung-sofef00-bonito.c} (57%)

diff --git a/drivers/gpu/drm/panel/Kconfig b/drivers/gpu/drm/panel/Kconfig
index d3a9a9f..d04f11f 100644
--- a/drivers/gpu/drm/panel/Kconfig
+++ b/drivers/gpu/drm/panel/Kconfig
@@ -741,6 +741,15 @@ config DRM_PANEL_SAMSUNG_SOFEF00
 
 	  The panels are 2280x1080@60Hz and 2340x1080@60Hz respectively
 
+config DRM_PANEL_SAMSUNG_SOFEF00_BONITO
+	tristate "Samsung sofef00 Google Pixel 3a XL DSI cmd mode panel"
+	depends on OF
+	depends on DRM_MIPI_DSI
+	depends on BACKLIGHT_CLASS_DEVICE
+	help
+	  Say Y or M here if you want to enable support for the Samsung AMOLED
+	  command mode panel found in the Pixel 3a XL smartphone.
+
 config DRM_PANEL_SEIKO_43WVF1G
 	tristate "Seiko 43WVF1G panel"
 	depends on OF
diff --git a/drivers/gpu/drm/panel/Makefile b/drivers/gpu/drm/panel/Makefile
index 987a0870..f30fd20 100644
--- a/drivers/gpu/drm/panel/Makefile
+++ b/drivers/gpu/drm/panel/Makefile
@@ -77,6 +77,7 @@ obj-$(CONFIG_DRM_PANEL_SAMSUNG_S6E63M0_DSI) += panel-samsung-s6e63m0-dsi.o
 obj-$(CONFIG_DRM_PANEL_SAMSUNG_S6E88A0_AMS452EF01) += panel-samsung-s6e88a0-ams452ef01.o
 obj-$(CONFIG_DRM_PANEL_SAMSUNG_S6E8AA0) += panel-samsung-s6e8aa0.o
 obj-$(CONFIG_DRM_PANEL_SAMSUNG_SOFEF00) += panel-samsung-sofef00.o
+obj-$(CONFIG_DRM_PANEL_SAMSUNG_SOFEF00_BONITO) += panel-samsung-sofef00-bonito.o
 obj-$(CONFIG_DRM_PANEL_SEIKO_43WVF1G) += panel-seiko-43wvf1g.o
 obj-$(CONFIG_DRM_PANEL_SHARP_LQ101R1SX01) += panel-sharp-lq101r1sx01.o
 obj-$(CONFIG_DRM_PANEL_SHARP_LS037V7DW01) += panel-sharp-ls037v7dw01.o
diff --git a/drivers/gpu/drm/panel/panel-samsung-s6e3fa7.c b/drivers/gpu/drm/panel/panel-samsung-sofef00-bonito.c
similarity index 57%
copy from drivers/gpu/drm/panel/panel-samsung-s6e3fa7.c
copy to drivers/gpu/drm/panel/panel-samsung-sofef00-bonito.c
index 10bc8fb..cb4c7e8 100644
--- a/drivers/gpu/drm/panel/panel-samsung-s6e3fa7.c
+++ b/drivers/gpu/drm/panel/panel-samsung-sofef00-bonito.c
@@ -1,10 +1,8 @@
 // SPDX-License-Identifier: GPL-2.0-only
 /*
- * Driver for the Samsung S6E3FA7 panel.
- *
- * Copyright (c) 2022-2024, The Linux Foundation. All rights reserved.
+ * Copyright (c) 2023 FIXME
  * Generated with linux-mdss-dsi-panel-driver-generator from vendor device tree:
- * Copyright (c) 2013, The Linux Foundation. All rights reserved.
+ * Copyright (c) 2013, The Linux Foundation. All rights reserved. (FIXME)
  */
 
 #include <linux/backlight.h>
@@ -19,18 +17,19 @@
 #include <drm/drm_modes.h>
 #include <drm/drm_panel.h>
 
-struct s6e3fa7_panel {
+struct sofef00 {
 	struct drm_panel panel;
 	struct mipi_dsi_device *dsi;
 	struct gpio_desc *reset_gpio;
+	bool prepared;
 };
 
-static inline struct s6e3fa7_panel *to_s6e3fa7_panel(struct drm_panel *panel)
+static inline struct sofef00 *to_sofef00(struct drm_panel *panel)
 {
-	return container_of(panel, struct s6e3fa7_panel, panel);
+	return container_of(panel, struct sofef00, panel);
 }
 
-static void s6e3fa7_panel_reset(struct s6e3fa7_panel *ctx)
+static void sofef00_reset(struct sofef00 *ctx)
 {
 	gpiod_set_value_cansleep(ctx->reset_gpio, 1);
 	usleep_range(1000, 2000);
@@ -38,7 +37,7 @@ static void s6e3fa7_panel_reset(struct s6e3fa7_panel *ctx)
 	usleep_range(10000, 11000);
 }
 
-static int s6e3fa7_panel_on(struct s6e3fa7_panel *ctx)
+static int sofef00_on(struct sofef00 *ctx)
 {
 	struct mipi_dsi_device *dsi = ctx->dsi;
 	struct device *dev = &dsi->dev;
@@ -49,7 +48,7 @@ static int s6e3fa7_panel_on(struct s6e3fa7_panel *ctx)
 		dev_err(dev, "Failed to exit sleep mode: %d\n", ret);
 		return ret;
 	}
-	msleep(120);
+	usleep_range(10000, 11000);
 
 	ret = mipi_dsi_dcs_set_tear_on(dsi, MIPI_DSI_DCS_TEAR_MODE_VBLANK);
 	if (ret < 0) {
@@ -57,12 +56,15 @@ static int s6e3fa7_panel_on(struct s6e3fa7_panel *ctx)
 		return ret;
 	}
 
-	mipi_dsi_dcs_write_seq(dsi, 0xf0, 0x5a, 0x5a);
-	mipi_dsi_dcs_write_seq(dsi, 0xf4,
-			       0xbb, 0x23, 0x19, 0x3a, 0x9f, 0x0f, 0x09, 0xc0,
-			       0x00, 0xb4, 0x37, 0x70, 0x79, 0x69);
+	ret = mipi_dsi_dcs_set_page_address(dsi, 0x0000, 0x086f);
+	if (ret < 0) {
+		dev_err(dev, "Failed to set page address: %d\n", ret);
+		return ret;
+	}
+
 	mipi_dsi_dcs_write_seq(dsi, 0xf0, 0xa5, 0xa5);
 	mipi_dsi_dcs_write_seq(dsi, MIPI_DCS_WRITE_CONTROL_DISPLAY, 0x20);
+	msleep(110);
 
 	ret = mipi_dsi_dcs_set_display_on(dsi);
 	if (ret < 0) {
@@ -73,76 +75,89 @@ static int s6e3fa7_panel_on(struct s6e3fa7_panel *ctx)
 	return 0;
 }
 
-static int s6e3fa7_panel_prepare(struct drm_panel *panel)
+static int sofef00_off(struct sofef00 *ctx)
 {
-	struct s6e3fa7_panel *ctx = to_s6e3fa7_panel(panel);
-	struct device *dev = &ctx->dsi->dev;
+	struct mipi_dsi_device *dsi = ctx->dsi;
+	struct device *dev = &dsi->dev;
 	int ret;
 
-	s6e3fa7_panel_reset(ctx);
+	ret = mipi_dsi_dcs_set_display_off(dsi);
+	if (ret < 0) {
+		dev_err(dev, "Failed to set display off: %d\n", ret);
+		return ret;
+	}
 
-	ret = s6e3fa7_panel_on(ctx);
+	ret = mipi_dsi_dcs_enter_sleep_mode(dsi);
 	if (ret < 0) {
-		dev_err(dev, "Failed to initialize panel: %d\n", ret);
-		gpiod_set_value_cansleep(ctx->reset_gpio, 1);
+		dev_err(dev, "Failed to enter sleep mode: %d\n", ret);
 		return ret;
 	}
+	msleep(120);
 
 	return 0;
 }
 
-static int s6e3fa7_panel_unprepare(struct drm_panel *panel)
+static int sofef00_prepare(struct drm_panel *panel)
 {
-	struct s6e3fa7_panel *ctx = to_s6e3fa7_panel(panel);
+	struct sofef00 *ctx = to_sofef00(panel);
+	struct device *dev = &ctx->dsi->dev;
+	int ret;
 
-	gpiod_set_value_cansleep(ctx->reset_gpio, 1);
+	if (ctx->prepared)
+		return 0;
+
+	sofef00_reset(ctx);
+
+	ret = sofef00_on(ctx);
+	if (ret < 0) {
+		dev_err(dev, "Failed to initialize panel: %d\n", ret);
+		gpiod_set_value_cansleep(ctx->reset_gpio, 1);
+		return ret;
+	}
 
+	ctx->prepared = true;
 	return 0;
 }
 
-static int s6e3fa7_panel_disable(struct drm_panel *panel)
+static int sofef00_unprepare(struct drm_panel *panel)
 {
-	struct s6e3fa7_panel *ctx = to_s6e3fa7_panel(panel);
-	struct mipi_dsi_device *dsi = ctx->dsi;
-	struct device *dev = &dsi->dev;
+	struct sofef00 *ctx = to_sofef00(panel);
+	struct device *dev = &ctx->dsi->dev;
 	int ret;
 
-	ret = mipi_dsi_dcs_set_display_off(dsi);
-	if (ret < 0) {
-		dev_err(dev, "Failed to set display off: %d\n", ret);
-		return ret;
-	}
+	if (!ctx->prepared)
+		return 0;
 
-	ret = mipi_dsi_dcs_enter_sleep_mode(dsi);
-	if (ret < 0) {
-		dev_err(dev, "Failed to enter sleep mode: %d\n", ret);
-		return ret;
-	}
-	msleep(120);
+	ret = sofef00_off(ctx);
+	if (ret < 0)
+		dev_err(dev, "Failed to un-initialize panel: %d\n", ret);
 
+	gpiod_set_value_cansleep(ctx->reset_gpio, 1);
+
+	ctx->prepared = false;
 	return 0;
 }
 
-static const struct drm_display_mode s6e3fa7_panel_mode = {
-	.clock = (1080 + 32 + 32 + 78) * (2220 + 32 + 4 + 78) * 60 / 1000,
+static const struct drm_display_mode sofef00_mode = {
+	.clock = (1080 + 32 + 32 + 98) * (2160 + 32 + 4 + 98) * 60 / 1000,
 	.hdisplay = 1080,
 	.hsync_start = 1080 + 32,
 	.hsync_end = 1080 + 32 + 32,
-	.htotal = 1080 + 32 + 32 + 78,
-	.vdisplay = 2220,
-	.vsync_start = 2220 + 32,
-	.vsync_end = 2220 + 32 + 4,
-	.vtotal = 2220 + 32 + 4 + 78,
-	.width_mm = 62,
-	.height_mm = 127,
+	.htotal = 1080 + 32 + 32 + 98,
+	.vdisplay = 2160,
+	.vsync_start = 2160 + 32,
+	.vsync_end = 2160 + 32 + 4,
+	.vtotal = 2160 + 32 + 4 + 98,
+	.width_mm = 69,
+	.height_mm = 137,
 };
 
-static int s6e3fa7_panel_get_modes(struct drm_panel *panel,
-				 struct drm_connector *connector)
+static int sofef00_get_modes(struct drm_panel *panel,
+			     struct drm_connector *connector)
 {
 	struct drm_display_mode *mode;
 
-	mode = drm_mode_duplicate(connector->dev, &s6e3fa7_panel_mode);
+	mode = drm_mode_duplicate(connector->dev, &sofef00_mode);
 	if (!mode)
 		return -ENOMEM;
 
@@ -156,46 +171,55 @@ static int s6e3fa7_panel_get_modes(struct drm_panel *panel,
 	return 1;
 }
 
-static const struct drm_panel_funcs s6e3fa7_panel_funcs = {
-	.prepare = s6e3fa7_panel_prepare,
-	.unprepare = s6e3fa7_panel_unprepare,
-	.disable = s6e3fa7_panel_disable,
-	.get_modes = s6e3fa7_panel_get_modes,
+static const struct drm_panel_funcs sofef00_panel_funcs = {
+	.prepare = sofef00_prepare,
+	.unprepare = sofef00_unprepare,
+	.get_modes = sofef00_get_modes,
 };
 
-static int s6e3fa7_panel_bl_update_status(struct backlight_device *bl)
+static int sofef00_bl_update_status(struct backlight_device *bl)
 {
 	struct mipi_dsi_device *dsi = bl_get_data(bl);
 	u16 brightness = backlight_get_brightness(bl);
 	int ret;
 
+	dsi->mode_flags &= ~MIPI_DSI_MODE_LPM;
+
 	ret = mipi_dsi_dcs_set_display_brightness_large(dsi, brightness);
 	if (ret < 0)
 		return ret;
 
+	dsi->mode_flags |= MIPI_DSI_MODE_LPM;
+
 	return 0;
 }
 
-static int s6e3fa7_panel_bl_get_brightness(struct backlight_device *bl)
+// TODO: Check if /sys/class/backlight/.../actual_brightness actually returns
+// correct values. If not, remove this function.
+static int sofef00_bl_get_brightness(struct backlight_device *bl)
 {
 	struct mipi_dsi_device *dsi = bl_get_data(bl);
 	u16 brightness;
 	int ret;
 
+	dsi->mode_flags &= ~MIPI_DSI_MODE_LPM;
+
 	ret = mipi_dsi_dcs_get_display_brightness_large(dsi, &brightness);
 	if (ret < 0)
 		return ret;
 
+	dsi->mode_flags |= MIPI_DSI_MODE_LPM;
+
 	return brightness;
 }
 
-static const struct backlight_ops s6e3fa7_panel_bl_ops = {
-	.update_status = s6e3fa7_panel_bl_update_status,
-	.get_brightness = s6e3fa7_panel_bl_get_brightness,
+static const struct backlight_ops sofef00_bl_ops = {
+	.update_status = sofef00_bl_update_status,
+	.get_brightness = sofef00_bl_get_brightness,
 };
 
 static struct backlight_device *
-s6e3fa7_panel_create_backlight(struct mipi_dsi_device *dsi)
+sofef00_create_backlight(struct mipi_dsi_device *dsi)
 {
 	struct device *dev = &dsi->dev;
 	const struct backlight_properties props = {
@@ -205,13 +229,13 @@ s6e3fa7_panel_create_backlight(struct mipi_dsi_device *dsi)
 	};
 
 	return devm_backlight_device_register(dev, dev_name(dev), dev, dsi,
-					      &s6e3fa7_panel_bl_ops, &props);
+					      &sofef00_bl_ops, &props);
 }
 
-static int s6e3fa7_panel_probe(struct mipi_dsi_device *dsi)
+static int sofef00_probe(struct mipi_dsi_device *dsi)
 {
 	struct device *dev = &dsi->dev;
-	struct s6e3fa7_panel *ctx;
+	struct sofef00 *ctx;
 	int ret;
 
 	ctx = devm_kzalloc(dev, sizeof(*ctx), GFP_KERNEL);
@@ -231,11 +255,10 @@ static int s6e3fa7_panel_probe(struct mipi_dsi_device *dsi)
 	dsi->mode_flags = MIPI_DSI_MODE_VIDEO_BURST |
 			  MIPI_DSI_CLOCK_NON_CONTINUOUS | MIPI_DSI_MODE_LPM;
 
-	drm_panel_init(&ctx->panel, dev, &s6e3fa7_panel_funcs,
+	drm_panel_init(&ctx->panel, dev, &sofef00_panel_funcs,
 		       DRM_MODE_CONNECTOR_DSI);
-	ctx->panel.prepare_prev_first = true;
 
-	ctx->panel.backlight = s6e3fa7_panel_create_backlight(dsi);
+	ctx->panel.backlight = sofef00_create_backlight(dsi);
 	if (IS_ERR(ctx->panel.backlight))
 		return dev_err_probe(dev, PTR_ERR(ctx->panel.backlight),
 				     "Failed to create backlight\n");
@@ -252,9 +275,9 @@ static int s6e3fa7_panel_probe(struct mipi_dsi_device *dsi)
 	return 0;
 }
 
-static void s6e3fa7_panel_remove(struct mipi_dsi_device *dsi)
+static void sofef00_remove(struct mipi_dsi_device *dsi)
 {
-	struct s6e3fa7_panel *ctx = mipi_dsi_get_drvdata(dsi);
+	struct sofef00 *ctx = mipi_dsi_get_drvdata(dsi);
 	int ret;
 
 	ret = mipi_dsi_detach(dsi);
@@ -264,22 +287,22 @@ static void s6e3fa7_panel_remove(struct mipi_dsi_device *dsi)
 	drm_panel_remove(&ctx->panel);
 }
 
-static const struct of_device_id s6e3fa7_panel_of_match[] = {
-	{ .compatible = "samsung,s6e3fa7-ams559nk06" },
+static const struct of_device_id sofef00_of_match[] = {
+	{ .compatible = "samsung,sofef00-bonito" }, // FIXME
 	{ /* sentinel */ }
 };
-MODULE_DEVICE_TABLE(of, s6e3fa7_panel_of_match);
+MODULE_DEVICE_TABLE(of, sofef00_of_match);
 
-static struct mipi_dsi_driver s6e3fa7_panel_driver = {
-	.probe = s6e3fa7_panel_probe,
-	.remove = s6e3fa7_panel_remove,
+static struct mipi_dsi_driver sofef00_driver = {
+	.probe = sofef00_probe,
+	.remove = sofef00_remove,
 	.driver = {
-		.name = "panel-samsung-s6e3fa7",
-		.of_match_table = s6e3fa7_panel_of_match,
+		.name = "panel-sofef00-bonito",
+		.of_match_table = sofef00_of_match,
 	},
 };
-module_mipi_dsi_driver(s6e3fa7_panel_driver);
+module_mipi_dsi_driver(sofef00_driver);
 
-MODULE_AUTHOR("Richard Acayan <mailingradian@gmail.com>");
-MODULE_DESCRIPTION("DRM driver for Samsung S6E3FA7 command mode DSI panel");
+MODULE_AUTHOR("linux-mdss-dsi-panel-driver-generator <fix@me>"); // FIXME
+MODULE_DESCRIPTION("DRM driver for sofef00 cmd mode dsi sdc panel");
 MODULE_LICENSE("GPL");
